frame/support/src/traits/hooks.rs

```diff
@@ -45,6 +47,29 @@ impl<BlockNumber: Clone> OnInitialize<BlockNumber> for Tuple {
 	}
 }
 
+/// The after inherent initialization trait.
+///
+/// Implementing this lets you express what should happen for your pallet when the block is
+/// beginning (right before the first extrinsic is executed).
+pub trait OnPostInherent<BlockNumber> {
+	/// We have finished executing all inherent extrinsics. Implement to have something happen.
+	///
+	/// Return the non-negotiable weight consumed for this behavior.
+	///
+	/// NOTE: Unlike `on_initialize`, this is called after the inherent extrinsics have been
+	/// included, so things like the `timestamp` will be up to date here.
+	fn on_post_inherent(_n: BlockNumber) -> crate::weights::Weight { 0 }
+}
+
+#[impl_for_tuples(30)]
+impl<BlockNumber: Clone> OnPostInherent<BlockNumber> for Tuple {
+	fn on_post_inherent(n: BlockNumber) -> crate::weights::Weight {
+		let mut weight = 0;
+		for_tuples!( #( weight = weight.saturating_add(Tuple::on_post_inherent(n.clone())); )* );
+		weight
+	}
+}
```

Comment:
We could also start using the name `on_pre_ingerent` instead of `on_initialize`. I think it is a better to understand this way, but needs a bit of boilerplate. 

```
pub trait OnPreInherent<BlockNumber> {
	fn on_pre_inherent(_n: BlockNumber) -> crate::weights::Weight { 0 }
}

#[impl_for_tuples(30)]
impl<BlockNumber: Clone> OnPreInherent<BlockNumber> for Tuple {
	fn on_pre_inherent(n: BlockNumber) -> crate::weights::Weight {
		let mut weight = 0;
		for_tuples!( #( weight = weight.saturating_add(Tuple::on_post_inherent(n.clone())); )* );
		weight
	}
}

// Any of the old OnInitialize implementation are also considered a valid implementation of 
// `OnPreInherent`
impl<BlockNumber, I: OnInitialize<BlockNumber>> OnPreInherent<BlockNumber> for I {
	fn on_pre_inherent(n: BlockNumber) -> crate::weights::Weight {
		Self::on_initialize(n)
	}
}
```